spec/models/article_spec.rb

```diff
@@ -136,6 +136,31 @@ def build_and_validate_article(*args)
       end
     end
 
+    describe "#canonical_url_must_be_unique" do
+      let(:url) { "http://example.com" }
+
+      it "is valid when the canonical url is nil" do
+        another_article = build(:article)
+
+        expect(another_article).to be_valid
+      end
+
+      it "is not valid when the url has been taken by a published article" do
+        article.update(canonical_url: url)
+        another_article = build(:article)
+        another_article.canonical_url = url
+
+        expect(another_article).not_to be_valid
+      end
+
+      it "is valid when the url has been taken by a draft article" do
+        another_article = build(:article)
+        another_article.canonical_url = url
+
+        expect(another_article).to be_valid
```

Comment:
I think there's a bug in this test now that I look at it closely. The article's still published and in memory. To create a draft article I think you have to use `create(:article, published: false)`

Reply:
You're right. I fixed the tests so that articles are created correctly, and the spec fails without the fix.